README.md
# Hourly Heatmap for Grafana

[![Build](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/workflows/CI/badge.svg)](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/actions?query=workflow%3A%22CI%22)
[![Release](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/workflows/Release/badge.svg)](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/actions?query=workflow%3ARelease)
[![Marketplace](https://img.shields.io/badge/dynamic/json?logo=grafana&color=F47A20&label=marketplace&prefix=v&query=%24.items%5B%3F%28%40.slug%20%3D%3D%20%22marcusolsson-hourly-heatmap-panel%22%29%5D.version&url=https%3A%2F%2Fgrafana.com%2Fapi%2Fplugins)](https://grafana.com/grafana/plugins/marcusolsson-hourly-heatmap-panel)
[![Downloads](https://img.shields.io/badge/dynamic/json?logo=grafana&color=F47A20&label=downloads&query=%24.items%5B%3F%28%40.slug%20%3D%3D%20%22marcusolsson-hourly-heatmap-panel%22%29%5D.downloads&url=https%3A%2F%2Fgrafana.com%2Fapi%2Fplugins)](https://grafana.com/grafana/plugins/marcusolsson-hourly-heatmap-panel)
[![License](https://img.shields.io/github/license/marcusolsson/grafana-hourly-heatmap-panel)](LICENSE)
[![Twitter](https://img.shields.io/twitter/follow/marcusolsson?color=%231DA1F2&label=twitter&style=plastic)](https://twitter.com/marcusolsson)
![Maintenance](https://img.shields.io/maintenance/no/2022?style=plastic)

> **Important:** As of July 2022, I'm no longer actively maintaining this plugin. At the time of writing, I'm not aware of any replacement panel.

A panel plugin for [Grafana](https://grafana.com) to visualize hourly heatmaps.

An hourly heatmap aggregates data into buckets by day and hour to analyze activity or traffic during the day.

![Screenshot](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/raw/main/src/img/dark.png)

## Motivation

The [carpet-plot](https://github.com/petrslavotinek/grafana-carpetplot) panel plugin is one of the most used plugins for Grafana. Unfortunately, it's no longer being actively maintained.

Grafana 7.0 introduced a new plugin architecture based on React. Instead of migrating the original plugin from Angular, this is completely rewritten from scratch, using inspiration from the original plugin.

## Configuration

This section lists the available configuration options for the JSON API data source.

### Panel options

#### Dimensions

| Option  | Description                                                             |
|---------|-------------------------------------------------------------------------|
| _Time_  | Name of the field to use for time. Defaults to the first time field.    |
| _Value_ | Name of the field to use for value. Defaults to the first number field. |

#### Display

| Option             | Description                                                                                                                                                                                                                                                                     |
|--------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| _Show cell border_ | Toggles a cell border to make it easier to distinguish cells with similar values                                                                                                                                                                                                |
| _Show tooltip_     | Toggles the tooltip. Due to the current tooltip implementation, this severely impacts performance and I recommend that you disable this for large time intervals. For more information, refer to [#12](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/issues/12). |
| _From_ and _To_    | Lets you choose the hours to display. This can be used to set working hours, or to filter parts of the day with low traffic                                                                                                                                                     |

#### Legend

| Option                 | Description                                                                                                                                                  |
|------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| _Show legend_          | Toggles the color spectrum                                                                                                                                   |
| _Show value indicator_ | Toggles an indicator that shows the current value in the legend                                                                                              |
| _Gradient quality_     | Determines the quality of the color spectrum. Higher quality means more SVG elements being drawn. Reduce the quality if you experience degraded performance. |

### Field options

| Option                 | Description                                                                                                                                                                                                          |
|------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| _Group by_             | Size of each bucket                                                                                                                                                                                                  |
| _Calculation_          | Calculation to use for reducing data within a bucket                                                                                                                                                                 |
| _Color palette_        | Colors to use for the heatmap. Select from any of the predefined color palettes, or select **Custom** to create your own. Select **Field options** to use the colors from the built-in **Color scheme** field option |
| _Invert color palette_ | Inverts the currently selected color palette                                                                                                                                                                         |
| _Null value color_     | Color to display when a bucket contains no data as opposed to a data value of 0.                                                                                                                                     |

## Troubleshooting

### Missing data

By default, data sources limits the number of data points to the width of the panel in pixels. If you're visualizing data over a long time, then you may need to adjust the **Max data points** under **Query options** in the query editor.

![Missing data](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/raw/main/src/img/missing-datapoints.png)


CHANGELOG.md
# Changelog

## 2.0.1 (2022-08-28)

[Full changelog](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/compare/v2.0.0...v2.0.1)

Update to Grafana 9.

## 2.0.0 (2021-11-19)

[Full changelog](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/compare/v1.0.0...v2.0.0)

### Enhancements

- Add option for 120 minute intervals ([#40](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/pull/40)) (thanks [@ChrizZz90](https://github.com/ChrizZz90)!)
- BREAKING CHANGE: Update to Grafana 8 theme API

## 1.0.0 (2021-06-15)

[Full changelog](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/compare/v0.10.0...v1.0.0)

### Enhancements

- Set color for null values ([#24](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/pull/24)) (thanks [@KatrinaTurner](https://github.com/KatrinaTurner)!)
- Fix dates in changelog ([#26](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/pull/26)) (thanks [@dnrce](https://github.com/dnrce)!)
- Fix display processor bug in Grafana 8

## 0.10.0 (2021-02-16)

[Full changelog](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/compare/v0.9.1...v0.10.0)

### Enhancements

- Make dimensions clearable
- Add fallback panel for unsupported Grafana versions
- Add wizard for configuring the query

## 0.9.1 (2021-01-13)

[Full changelog](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/compare/v0.9.0...v0.9.1)

### Bug fixes

- Min and max were incorrectly calculated for aggregations ([#16](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/issues/16))

## 0.9.0 (2020-12-07)

[Full changelog](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/compare/v0.8.1...v0.9.0)

### Enhancements

- Highlight the legend section for the selected hour ([#15](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/issues/15))
- Select the fields to use for time and value
- Add option to configure legend quality
- Add option to use the new color scheme field option in Grafana 7.3+
- Add option to show a cell border ([#17](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/issues/17))
- Add option to disable tooltips

## 0.8.1 (2020-11-27)

[Full changelog](https://github.com/marcusolsson/grafana-hourly-heatmap-panel/compare/v0.8.0...v0.8.1)

### Enhancements

- Updated `@grafana` dependencies from `^7.0.0` to `^7.3.0`
- Improved release process using the new [GitHub workflows](https://github.com/grafana/plugin-workflows) for Grafana plugins


src/bucket.test.ts
describe('bucketize', () => {
  it('single row', () => {
    expect(true).toBeTruthy();
  });
});
// import { ArrayVector, dateTime, dateTimeParse, Field, FieldType, TimeRange } from '@grafana/data';
// import { bucketize, defaultDisplay } from './bucket';

// const timeZone = 'utc';
// const startTime = dateTimeParse('2020-05-05T00:00:00Z', { timeZone });
// const dailyInterval: [number, number] = [0, 24];
// const timeRange: TimeRange = {
//   from: startTime,
//   to: dateTime(startTime).add(1, 'day'),
//   raw: {
//     from: `${1588683600000}`,
//     to: `${dateTime(startTime)
//       .add(1, 'day')
//       .valueOf()}`,
//   },
// };

// describe('bucketize', () => {
//   it('single row', () => {
//     const timeField: Field = {
//       name: 'time',
//       type: FieldType.time,
//       config: {},
//       values: new ArrayVector([
//         dateTime(startTime)
//           .add(13, 'hours')
//           .add(20, 'minutes'),
//       ]),
//     };

//     const valueField: Field = {
//       name: 'value',
//       type: FieldType.number,
//       config: {
//         min: 0,
//         max: 1,
//         custom: {
//           groupBy: 60,
//           calculation: 'sum',
//         },
//       },
//       values: new ArrayVector([1]),
//     };

//     expect(bucketize(timeField, valueField, timeZone, timeRange, dailyInterval)).toStrictEqual({
//       min: 0,
//       max: 1,
//       numBuckets: 24,
//       points: [
//         {
//           bucketStartMillis: dateTime(startTime)
//             .add(13, 'hours')
//             .valueOf(),
//           dayMillis: 1588683600000,
//           value: 1,
//         },
//       ],
//       timeDisplay: defaultDisplay,
//       valueDisplay: defaultDisplay,
//     });
//   });

//   it('sum bucket data', () => {
//     const timeField: Field = {
//       name: 'time',
//       type: FieldType.time,
//       config: {},
//       values: new ArrayVector(
//         [
//           // First hour
//           dateTime(startTime)
//             .add(13, 'hours')
//             .add(20, 'minutes'),
//           dateTime(startTime)
//             .add(13, 'hours')
//             .add(30, 'minutes'),
//           dateTime(startTime)
//             .add(13, 'hours')
//             .add(40, 'minutes'),

//           // Second hour
//           dateTime(startTime)
//             .add(14, 'hours')
//             .add(20, 'minutes'),
//           dateTime(startTime)
//             .add(14, 'hours')
//             .add(30, 'minutes'),
//           dateTime(startTime)
//             .add(14, 'hours')
//             .add(40, 'minutes'),

//           // Third hour
//           dateTime(startTime)
//             .add(15, 'hours')
//             .add(20, 'minutes'),
//           dateTime(startTime)
//             .add(15, 'hours')
//             .add(30, 'minutes'),
//           dateTime(startTime)
//             .add(15, 'hours')
//             .add(40, 'minutes'),
//         ].map(_ => _.valueOf())
//       ),
//     };

//     const valueField: Field = {
//       name: 'value',
//       type: FieldType.number,
//       config: {
//         min: 0,
//         max: 10,
//         custom: {
//           groupBy: 60,
//           calculation: 'sum',
//         },
//       },
//       values: new ArrayVector([
//         // First hour
//         5,
//         5,
//         5,

//         // Second hour
//         7,
//         3,
//         10,

//         // Third hour

//         0,
//         0,
//         0,
//       ]),
//     };

//     expect(bucketize(timeField, valueField, timeZone, timeRange, dailyInterval)).toStrictEqual({
//       min: 0,
//       max: 10,
//       numBuckets: 24,
//       points: [
//         {
//           bucketStartMillis: dateTime(startTime)
//             .add(13, 'hours')
//             .valueOf(),
//           dayMillis: 1588683600000,
//           value: 15,
//         },
//         {
//           bucketStartMillis: dateTime(startTime)
//             .add(14, 'hours')
//             .valueOf(),
//           dayMillis: 1588687200000,
//           value: 20,
//         },
//         {
//           bucketStartMillis: dateTime(startTime)
//             .add(15, 'hours')
//             .valueOf(),
//           dayMillis: 1588690800000,
//           value: 0,
//         },
//       ],
//       timeDisplay: defaultDisplay,
//       valueDisplay: defaultDisplay,
//     });
//   });

//   it('average bucket data', () => {
//     const timeField: Field = {
//       name: 'time',
//       type: FieldType.time,
//       config: {},
//       values: new ArrayVector(
//         [
//           // First hour
//           dateTime(startTime)
//             .add(13, 'hours')
//             .add(20, 'minutes'),
//           dateTime(startTime)
//             .add(13, 'hours')
//             .add(30, 'minutes'),
//           dateTime(startTime)
//             .add(13, 'hours')
//             .add(40, 'minutes'),

//           // Second hour
//           dateTime(startTime)
//             .add(14, 'hours')
//             .add(20, 'minutes'),
//           dateTime(startTime)
//             .add(14, 'hours')
//             .add(30, 'minutes'),
//           dateTime(startTime)
//             .add(14, 'hours')
//             .add(40, 'minutes'),

//           // Third hour
//           dateTime(startTime)
//             .add(15, 'hours')
//             .add(20, 'minutes'),
//           dateTime(startTime)
//             .add(15, 'hours')
//             .add(30, 'minutes'),
//           dateTime(startTime)
//             .add(15, 'hours')
//             .add(40, 'minutes'),
//         ].map(_ => _.valueOf())
//       ),
//     };

//     const valueField: Field = {
//       name: 'value',
//       type: FieldType.number,
//       config: {
//         min: 0,
//         max: 10,
//         custom: {
//           groupBy: 60,
//           calculation: 'mean',
//         },
//       },
//       values: new ArrayVector([
//         // First hour
//         5,
//         5,
//         5,

//         // Second hour
//         7,
//         3,
//         10,

//         // Third hour

//         0,
//         0,
//         0,
//       ]),
//     };

//     expect(bucketize(timeField, valueField, timeZone, timeRange, dailyInterval)).toStrictEqual({
//       min: 0,
//       max: 10,
//       numBuckets: 24,
//       points: [
//         {
//           bucketStartMillis: dateTime(startTime)
//             .add(13, 'hours')
//             .valueOf(),
//           dayMillis: 1588683600000,
//           value: 5,
//         },
//         {
//           bucketStartMillis: dateTime(startTime)
//             .add(14, 'hours')
//             .valueOf(),
//           dayMillis: 1588687200000,
//           value: 6.666666666666667,
//         },
//         {
//           bucketStartMillis: dateTime(startTime)
//             .add(15, 'hours')
//             .valueOf(),
//           dayMillis: 1588690800000,
//           value: 0,
//         },
//       ],
//       timeDisplay: defaultDisplay,
//       valueDisplay: defaultDisplay,
//     });
//   });
// });


src/types.ts
import { ThresholdsConfig } from '@grafana/data';
import { TimeRegion } from './components/TimeRegionEditor';

export type Quality = 'low' | 'medium' | 'high';
export type Calculation = 'mean' | 'sum' | 'count' | 'min' | 'max' | 'first' | 'last';

export interface HeatmapOptions {
  // Dimensions
  timeFieldName?: string;
  valueFieldName?: string;

  // Display
  from: string;
  to: string;
  regions: TimeRegion[];
  showCellBorder: boolean;
  showTooltip: boolean;

  // Legend
  showLegend: boolean;
  showValueIndicator: boolean;
  legendGradientQuality: Quality;
}

export interface HeatmapFieldConfig {
  colorPalette: string;
  invertPalette: boolean;
  nullValueColor: string;

  // Options for custom color palettes.
  colorSpace: string;
  thresholds: ThresholdsConfig;

  // Options for reducing buckets.
  calculation: Calculation;
  groupBy: number;
}


src/bucket.ts
import {
  dateTime,
  dateTimeParse,
  DisplayProcessor,
  Field,
  getDisplayProcessor,
  GrafanaTheme2,
  TimeRange,
  TimeZone,
} from '@grafana/data';
import * as d3 from 'd3';

export interface Point {
  time: number;
  value: number;
}

interface PointSet {
  time: number;
  values: Point[];
}

export interface BucketPoint {
  // dayMillis corresponds to the start of the day.
  dayMillis: number;

  // bucketStartMillis corresponds to the start of a bucket, for example the start of every hour.
  bucketStartMillis: number;

  value: number;
}

export interface BucketData {
  numBuckets: number;
  points: BucketPoint[];
  valueField: Field<number>;
  timeField: Field<number>;
}

// reduce applies a calculation to an aggregated point set.
const reduce = (agg: PointSet[], calculation: (n: Array<number | undefined>) => number): Point[] => {
  return agg.map(({ time, values }) => ({
    time: time,
    value: calculation(values.map(({ value }) => value)),
  }));
};

// groupByMinutes groups a set of points
export const groupByMinutes = (points: Point[], minutes: number, timeZone: string): PointSet[] => {
  // Create keys for interval start.
  const rounded = points.map((point) => {
    const intervalStart = dateTime(point.time);
    intervalStart.subtract(intervalStart.minute ? intervalStart.minute() % minutes : 0.0, 'minutes');

    return {
      [intervalStart.valueOf()]: point,
    };
  });

  // Group by interval start.
  return Object.entries(
    rounded.reduce((acc: any, val: any) => {
      const intervalStartMillis: number = parseFloat(Object.keys(val)[0]);
      (acc[intervalStartMillis] = acc[intervalStartMillis] || []).push(val[intervalStartMillis]);
      return acc;
    }, {})
  ).map(([key, values]) => ({
    time: parseFloat(key),
    values: values as Point[],
  }));
};

// groupByDay works like to groupByDays but is a bit simpler.
export const groupByDay = (points: Point[]): PointSet[] => {
  const rounded = points.map((point) => ({
    [dateTime(point.time).startOf('day').valueOf()]: point,
  }));

  return Object.entries(
    rounded.reduce((acc: any, val: any) => {
      const dayMillis: number = parseFloat(Object.keys(val)[0]);
      (acc[dayMillis] = acc[dayMillis] || []).push(val[dayMillis]);
      return acc;
    }, {})
  ).map(([key, values]) => ({
    time: parseFloat(key),
    values: values as Point[],
  }));
};

export const defaultDisplay: DisplayProcessor = (value: any) => ({ numeric: value, text: value.toString() });
const minutesPerDay = 24 * 60;

// bucketize returns the main data structure used by the visualizations.
export const bucketize = (
  timeField: Field<number>,
  valueField: Field<number>,
  timeZone: string,
  timeRange: TimeRange,
  dailyInterval: [number, number],
  theme: GrafanaTheme2
): BucketData => {
  // Convert data frame fields to rows.
  const rows = Array.from({ length: timeField.values.length }, (_, i) => ({
    time: timeField.values.get(i),
    value: valueField.values.get(i),
  }));

  // Get the time range extents in the dashboard time zone.
  const extents = [
    dateTimeParse(timeRange.from.valueOf(), { timeZone }).startOf('day'),
    dateTimeParse(timeRange.to.valueOf(), { timeZone }).endOf('day'),
  ];

  const rowsWithinTimeRange = rows.filter((row) => {
    // Filter points within time range.
    const curr = dateTimeParse(row.time, { timeZone });
    return extents[0].valueOf() <= curr.valueOf() && curr.valueOf() < extents[1].valueOf();
  });

  const rowsWithinDailyInterval = rowsWithinTimeRange.filter((row) => {
    // Filter rows within interval.
    const dt = dateTimeParse(row.time, { timeZone });
    const hour = dt.hour ? dt.hour() : 0.0;
    return dailyInterval[0] <= hour && hour < dailyInterval[1];
  });

  // Extract the field configuration..
  const customData = valueField.config.custom;

  // Group and reduce values.
  const groupedByMinutes = groupByMinutes(rowsWithinDailyInterval, customData.groupBy, timeZone);
  const reducedMinutes = reduce(groupedByMinutes, calculations[customData.calculation]);

  const aggregatedPoints = groupByDay(reducedMinutes).flatMap(({ time, values }) =>
    values.map(({ time, value }) => ({
      dayMillis: time,
      bucketStartMillis: time,
      value,
    }))
  );

  recalculateMinMax(
    valueField,
    aggregatedPoints.map(({ value }) => value),
    theme,
    timeZone
  );

  return {
    numBuckets: Math.floor(minutesPerDay / customData.groupBy),
    points: aggregatedPoints,
    valueField,
    timeField,
  };
};

// recalculateMinMax updates the field min and max to the extents of the
// aggregated values rather than the raw values.
//
// TODO: While this works, it feels like hacky. Is there a better way to do this?
const recalculateMinMax = (field: Field<number>, values: number[], theme: GrafanaTheme2, timeZone: TimeZone) => {
  // Future versions of Grafana will change how the min and max are calculated.
  // For example, if Min or Max are set to auto, they will be undefined.
  //
  // Also, we should probably use getFieldConfigWithMinMax in the future:
  // https://github.com/grafana/grafana/blob/097dcc456a617bc67c3d5134e22adc00ad5b79c5/packages/grafana-data/src/field/scale.ts#L68
  const autoMin = !field.config.min || field.config.min === field.state?.calcs?.min;
  const autoMax = !field.config.max || field.config.max === field.state?.calcs?.max;

  if (autoMin) {
    field.config.min = d3.min(values);
  }
  if (autoMax) {
    field.config.max = d3.max(values);
  }

  field.display = getDisplayProcessor({
    field,
    theme,
    timeZone,
  });
};

// Lookup table for calculations.
const calculations: any = {
  mean: (vals: number[]) => d3.mean(vals),
  sum: (vals: number[]) => d3.sum(vals),
  count: (vals: number[]) => vals.length,
  min: (vals: number[]) => d3.min(vals),
  max: (vals: number[]) => d3.max(vals),
  first: (vals: number[]) => vals[0],
  last: (vals: number[]) => vals[vals.length - 1],
};


src/colors.ts
import { ThresholdsConfig, ThresholdsMode } from '@grafana/data';
import * as d3 from 'd3';

type Interpolator = (t: number) => string;

interface ColorScaleLookup {
  [name: string]: Interpolator;
}

const interpolators: ColorScaleLookup = {
  // Diverging
  interpolateSpectral: d3.interpolateSpectral,
  interpolateRdYlGn: d3.interpolateRdYlGn,

  // Single hue
  interpolateBlues: d3.interpolateBlues,
  interpolateGreens: d3.interpolateGreens,
  interpolateGreys: d3.interpolateGreys,
  interpolateOranges: d3.interpolateOranges,
  interpolatePurples: d3.interpolatePurples,
  interpolateReds: d3.interpolateReds,

  // Multi-hue
  interpolateBuGn: d3.interpolateBuGn,
  interpolateBuPu: d3.interpolateBuPu,
  interpolateGnBu: d3.interpolateGnBu,
  interpolateOrRd: d3.interpolateOrRd,
  interpolatePuBuGn: d3.interpolatePuBuGn,
  interpolatePuBu: d3.interpolatePuBu,
  interpolatePuRd: d3.interpolatePuRd,
  interpolateRdPu: d3.interpolateRdPu,
  interpolateYlGnBu: d3.interpolateYlGnBu,
  interpolateYlGn: d3.interpolateYlGn,
  interpolateYlOrBr: d3.interpolateYlOrBr,
  interpolateYlOrRd: d3.interpolateYlOrRd,
};

// makeSpectrumColorScale returns a sequential scale that maps numbers between
// min and max, using a given color palette.
export const makeSpectrumColorScale = (
  palette: string,
  min: number,
  max: number,
  invertPalette: boolean
): d3.ScaleSequential<string> => {
  if (invertPalette) {
    return d3.scaleSequential(interpolators[palette]).domain([max, min]);
  } else {
    return d3.scaleSequential(interpolators[palette]).domain([min, max]);
  }
};

interface InterpolatorLookup {
  [name: string]: d3.InterpolatorFactory<string, string>;
}

const interpolationMap: InterpolatorLookup = {
  rgb: d3.interpolateRgb,
  hsl: d3.interpolateHsl,
  hcl: d3.interpolateHcl,
  lab: d3.interpolateLab,
  cubehelix: d3.interpolateCubehelix,
};

// makeCustomColorScale returns a linear scale that maps numbers between min and
// max. Instead of using a predefined D3 interpolator, the color palette is
// created from a thresholds configuration.
export const makeCustomColorScale = (colorSpace: string, min: number, max: number, thresholds: ThresholdsConfig) => {
  const interpolator = interpolationMap[colorSpace];
  const colorRange = thresholds.steps.map(({ color }) => color);

  if (thresholds.mode === ThresholdsMode.Absolute) {
    return d3
      .scaleLinear<string>()
      .domain(thresholds.steps.map(({ value }) => (value >= 0 ? value : 0.0)))
      .range(colorRange)
      .interpolate(interpolator);
  }

  // If the thresholds mode is `percentage`, we first need to map the percentage
  // from [0, 100] to [min, max].
  const scale = d3.scaleLinear().domain([0, 100]).range([min, max]);

  return d3
    .scaleLinear<string>()
    .domain(thresholds.steps.map(({ value }): number => (value >= 0 ? scale(value) ?? 0.0 : 0.0)))
    .range(colorRange)
    .interpolate(interpolator);
};


src/module.ts
import {
  dateTime,
  FieldConfigProperty,
  FieldType,
  PanelPlugin,
  standardEditorsRegistry,
  thresholdsOverrideProcessor,
} from '@grafana/data';
import * as d3 from 'd3';
import {
  FieldSelectEditor,
  getPanelPluginOrFallback,
  hasCapability,
  standardOptionsCompat,
} from 'grafana-plugin-support';
import { TimeRegionEditor } from './components/TimeRegionEditor';
import { HeatmapPanel } from './HeatmapPanel';
import { HeatmapFieldConfig, HeatmapOptions } from './types';

const paletteSelected = (colorPalette: string) => (config: HeatmapFieldConfig) => config.colorPalette === colorPalette;

const buildStandardOptions = (): any => {
  const options = [
    FieldConfigProperty.Min,
    FieldConfigProperty.Max,
    FieldConfigProperty.Decimals,
    FieldConfigProperty.Unit,
  ];

  if (hasCapability('color-scheme')) {
    options.push(FieldConfigProperty.Color);
    options.push(FieldConfigProperty.Thresholds);
  }

  return standardOptionsCompat(options);
};

const buildColorPaletteOptions = () => {
  const options: Array<{ label: string; value: string; description?: string }> = [
    { value: 'custom', label: 'Custom', description: 'Define a custom color palette' },
  ];

  if (hasCapability('color-scheme')) {
    options.push({
      value: 'fieldOptions',
      label: 'Field options',
      description: 'Use color from field options (available from Grafana 7.3)',
    });
  }

  return options.concat([...predefinedColorPalettes]);
};

export const plugin = getPanelPluginOrFallback(
  'marcusolsson-hourly-heatmap-panel',
  new PanelPlugin<HeatmapOptions, HeatmapFieldConfig>(HeatmapPanel)
    .useFieldConfig({
      useCustomConfig: (builder) => {
        builder
          .addSelect({
            path: 'groupBy',
            name: 'Group by',
            settings: {
              options: [
                { value: 15, label: '15 minutes' },
                { value: 30, label: '30 minutes' },
                { value: 60, label: '60 minutes' },
                { value: 120, label: '120 minutes' },
              ],
            },
            defaultValue: 60,
          })
          .addSelect({
            path: 'calculation',
            name: 'Calculation',
            settings: {
              options: [
                { value: 'mean', label: 'Mean' },
                { value: 'sum', label: 'Sum' },
                { value: 'count', label: 'Count' },
                { value: 'min', label: 'Min' },
                { value: 'max', label: 'Max' },
                { value: 'first', label: 'First' },
                { value: 'last', label: 'Last' },
              ],
            },
            defaultValue: 'mean',
          })
          .addSelect({
            path: 'colorPalette',
            name: 'Color palette',
            settings: {
              options: buildColorPaletteOptions(),
            },
            defaultValue: 'interpolateSpectral',
          })
          .addBooleanSwitch({
            path: 'invertPalette',
            name: 'Invert color palette',
            defaultValue: false,
            showIf: (config: HeatmapFieldConfig) =>
              config.colorPalette !== 'custom' && config.colorPalette !== 'fieldOptions',
          })
          .addColorPicker({
            path: 'nullValueColor',
            name: 'Null value color',
            defaultValue: 'rgb(155, 155, 155)',
          })
          .addSelect({
            path: 'colorSpace',
            name: 'Color space',
            settings: {
              options: [
                { value: 'rgb', label: 'RGB' },
                { value: 'hsl', label: 'HSL' },
                { value: 'hcl', label: 'HCL' },
                { value: 'lab', label: 'Lab' },
                { value: 'cubehelix', label: 'Cubehelix' },
              ],
            },
            showIf: paletteSelected('custom'),
            defaultValue: 'rgb',
          })
          .addCustomEditor({
            id: 'thresholds',
            path: 'thresholds',
            name: 'Thresholds',
            editor: standardEditorsRegistry.get('thresholds').editor as any,
            override: standardEditorsRegistry.get('thresholds').editor as any,
            process: thresholdsOverrideProcessor,
            shouldApply: (field) => field.type === FieldType.number,
            showIf: paletteSelected('custom'),
          });
      },
      standardOptions: buildStandardOptions(),
    })
    .setPanelOptions((builder) => {
      return builder
        .addCustomEditor({
          id: 'timeFieldName',
          path: 'timeFieldName',
          name: 'Time',
          description: 'Defaults to the first time field.',
          category: ['Dimensions'],
          editor: FieldSelectEditor,
          settings: {
            filterByType: [FieldType.time],
          },
        })
        .addCustomEditor({
          id: 'valueFieldName',
          path: 'valueFieldName',
          name: 'Value',
          description: 'Defaults to the first number field.',
          category: ['Dimensions'],
          editor: FieldSelectEditor,
          settings: {
            filterByType: [FieldType.number],
          },
        })
        .addBooleanSwitch({
          path: 'showCellBorder',
          name: 'Show cell border',
          description: 'Display a border around each cell',
          defaultValue: false,
        })
        .addBooleanSwitch({
          path: 'showTooltip',
          name: 'Show tooltip',
          description:
            "Display a tooltip for the value under the cursor. Currently, this severely impacts performance. It's recommended to disable this for large time intervals.",
          defaultValue: true,
        })
        .addSelect({
          path: 'from',
          name: 'From',
          description: 'Hide values before this hour',
          settings: {
            options: d3.range(0, 24, 1).map((h) => ({
              label: dateTime().startOf('day').add(h, 'hour').format('HH:mm'),
              value: `${h}`,
            })),
          },
          defaultValue: '0',
        })
        .addSelect({
          path: 'to',
          name: 'To',
          description: 'Hide values after this hour',
          settings: {
            options: d3.range(0, 24, 1).map((h) => ({
              label: dateTime().startOf('day').add(h, 'hour').format('HH:mm'),
              value: `${h}`,
            })),
          },
          defaultValue: '0',
        })
        .addCustomEditor({
          id: 'regions',
          path: 'regions',
          name: 'Time regions',
          description: 'Highlight time regions during the day',
          editor: TimeRegionEditor,
        })
        .addBooleanSwitch({
          path: 'showLegend',
          name: 'Show legend',
          defaultValue: true,
          category: ['Legend'],
        })
        .addBooleanSwitch({
          path: 'showValueIndicator',
          name: 'Show value indicator',
          description: 'Displays an indicator for the value under the cursor',
          defaultValue: false,
          category: ['Legend'],
        })
        .addSelect({
          path: 'legendGradientQuality',
          name: 'Gradient quality',
          description: 'Higher quality means more elements on screen. Reduce quality if the panel is slow.',
          defaultValue: 'high',
          category: ['Legend'],
          settings: {
            options: [
              { label: 'High', value: 'high' },
              { label: 'Medium', value: 'medium' },
              { label: 'Low', value: 'low' },
            ],
          },
        });
    })
);

const predefinedColorPalettes = [
  // Diverging
  { label: 'Spectral', value: 'interpolateSpectral' },
  { label: 'RdYlGn', value: 'interpolateRdYlGn' },

  // Sequential (Single Hue)
  { label: 'Blues', value: 'interpolateBlues' },
  { label: 'Greens', value: 'interpolateGreens' },
  { label: 'Greys', value: 'interpolateGreys' },
  { label: 'Oranges', value: 'interpolateOranges' },
  { label: 'Purples', value: 'interpolatePurples' },
  { label: 'Reds', value: 'interpolateReds' },

  // Sequential (Multi-Hue)
  { label: 'BuGn', value: 'interpolateBuGn' },
  { label: 'BuPu', value: 'interpolateBuPu' },
  { label: 'GnBu', value: 'interpolateGnBu' },
  { label: 'OrRd', value: 'interpolateOrRd' },
  { label: 'PuBuGn', value: 'interpolatePuBuGn' },
  { label: 'PuBu', value: 'interpolatePuBu' },
  { label: 'PuRd', value: 'interpolatePuRd' },
  { label: 'RdPu', value: 'interpolateRdPu' },
  { label: 'YlGnBu', value: 'interpolateYlGnBu' },
  { label: 'YlGn', value: 'interpolateYlGn' },
  { label: 'YlOrBr', value: 'interpolateYlOrBr' },
  { label: 'YlOrRd', value: 'interpolateYlOrRd' },
];
